daemo/client.py

```diff
@@ -1,51 +1,264 @@
+
+        if self._credentials_exist():
+            self._load_tokens()
+        else:
+            self._persist_tokens()
 
-class DaemoClient:
-    def __init__(self, host, access_token):
-        self.host = host
-        self.access_token = access_token
         self.session = requests.session()
-        headers = {'Authorization': 'Bearer {}'.format(self.access_token)}
-        factory = WebSocketClientFactory('ws://' + self.host + '/ws/api-client?subscribe-user', headers=headers)
-        factory.protocol = DaemoClientProtocol
-        connectWS(factory)
+        self.authenticate()
+
+    def authenticate(self):
+        self._refresh_token()
+
+    def publish(self, project_id, approve, completed, stream):
+        assert project_id is not None and project_id > 0, Error.required(PROJECT_ID)
+        assert isfunction(approve), Error.func_def_undefined(APPROVE)
+        assert isfunction(completed), Error.func_def_undefined(CALLBACK)
+        assert stream is not None, Error.required(STREAM)
+
+        self.project_id = project_id
+        self.stream = stream
+
+        self._launch(project_id, approve, completed, stream)
+
+    def add_data(self, project_id, data):
+        response = self._post('/api/project/%d/add-data/' % project_id, data=json.dumps(data))
+        return response
+
+    def fetch_task(self, taskworker_id):
+        response = self._get('/api/task-worker/%d/' % taskworker_id, data={})
+        return response
+
+    def update_status(self, task):
+        data = {
+            'status': STATUS_ACCEPTED if task['accept'] else STATUS_REJECTED,
+            'workers': [task['id']]
+        }
+
+        response = self._post('/api/task-worker/bulk-update-status/', data=json.dumps(data))
+        return response
+
+    def fetch_status(self, project_id):
+        response = self._post('/api/project/%d/is-done/' % project_id)
+        return response
+
+    def is_auth_error(self, response):
+        try:
+            response = response.json()
+        except Exception as e:
+            pass
+
+        auth_error = CREDENTIALS_NOT_PROVIDED
+
+        return response is not None and isinstance(response, dict) and response.get("detail", "") == auth_error
+
+    def _credentials_exist(self):
+        import os
+        return os.path.isfile(CREDENTIALS)
+
+    def _load_tokens(self):
+        import json
+        with open(self.credentials_path, READ_ONLY) as infile:
+            data = json.load(infile)
+
+            assert data[CLIENT_ID] is not None and len(data[CLIENT_ID]) > 0, Error.required(CLIENT_ID)
+            assert data[ACCESS_TOKEN] is not None and len(data[ACCESS_TOKEN]) > 0, Error.required(ACCESS_TOKEN)
+            assert data[REFRESH_TOKEN] is not None and len(data[REFRESH_TOKEN]) > 0, Error.required(REFRESH_TOKEN)
+
+            self.client_id = data[CLIENT_ID]
+            self.access_token = data[ACCESS_TOKEN]
+            self.refresh_token = data[REFRESH_TOKEN]
+        infile.close()
```

Comment:
when you use `with` the file is already closed properly when you leave that scope